lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php

```diff
@@ -24,6 +24,8 @@
 
 /**
  * Persister for many-to-many collections.
+ *
+ * @psalm-import-type AssociationMapping from ClassMetadata
```

Comment:
The most Psalm errors occur in this file, because Psalm does not know that `joinTable` is always set for mappings that represent the owning side of a many to many relationship.

As it stands, https://github.com/doctrine/orm/pull/10405 won't fix that either, because I did create a separate type for owning sides, but maybe I should?

## Before

```mermaid
classDiagram
    AssociationMapping <|-- OneToOneAssociationMapping
    AssociationMapping <|-- OneToManyAssociationMapping
    AssociationMapping <|-- ManyToOneAssociationMapping
    AssociationMapping <|-- ManyToManyAssociationMapping
```

## After

```mermaid
classDiagram
    AssociationMapping <|-- OneToOneAssociationMapping
    AssociationMapping <|-- OneToManyAssociationMapping
    AssociationMapping <|-- ManyToOneAssociationMapping
    AssociationMapping <|-- ManyToManyAssociationMapping
    ManyToManyAssociationMapping <|-- OwningManyToManyAssociationMapping
    ManyToManyAssociationMapping <|-- OwnedManyToManyAssociationMapping
    OneToOneAssociationMapping <|-- OwningOneToOneAssociationMapping
    OneToOneAssociationMapping <|-- OwnedOneToOneAssociationMapping
```

Cc @mpdude

Reply:
How could/would Psalm know when/where you're dealing with an `AssociationMapping` that represents the owning side? Would have have to make `instanceof` checks?

Reply:
Yes, exactly! We would replace the guard check here, for instance:

https://github.com/doctrine/orm/blob/31ff969628946ef2df87d0b6d49dabb1cc40119d/lib/Doctrine/ORM/Persisters/Collection/ManyToManyPersister.php#L37-L46

Reply:
Then let's see that we define the class hierarchy not only by technical criteria (i.e. which fields are common in several classes and extract those into a base class), but from a semantical point of view. What kind of checks do we have in the codebase?

For example `if ($mapping['isOwningSide']) ...` would become `if ($mapping instanceof AssociationOwningSide) ...` or so.

Reply:
I think we have two kind of checks: you just mentioned the first one, and the other one is against `$mapping['type']`, sometimes with the `&` operator, for instance to detect if it's a `to-many`, for instance here:
https://github.com/doctrine/orm/blob/31ff969628946ef2df87d0b6d49dabb1cc40119d/lib/Doctrine/ORM/PersistentCollection.php#L354

We might need to introduce `ToOneAssociationMapping` and `ToManyAssociationMapping` interfaces.

Reply:
Ah but we only ever test for `TO_MANY` and `TO_ONE`, never for `ONE_TO` or `MANY_TO` (those constants don't even exist), so it should be possible to just include these as classes in the hierarchy. I will ditch the traits and do that instead.

Reply:
I pushed a new commit on the other PR, switching from traits to classes.

> The thing is, interfaces can't have properties, so it's only a viable option if we switch to methods instead.

Just realised that this might not be an issue because we should be able to change the phpdoc on `ClassMetadata::AssociationMapping` from `array<string, AssociationMapping>` to `array<string, list-of-concrete-classes>`. An `instanceof` check should allow SA to rule out some of the classes and to notice that some of them have common properties :crossed_fingers: